enhancements/kube-apiserver/startup-monitor.md

```diff
@@ -0,0 +1,234 @@
+  When going back 2 or more revisions this invariant breaks. But it is the best we can do as the alternative is be unavailable.
+
+  Note: in etcd encryption controllers we wait until deployments settle. We won't make progress when the operator is in Progression=True state.
+
+### Open Questions [optional]
+
+### Test Plan
+
+The usual e2e tests will verify the happy case.
+
+For the error case, we have to inject the different types of errors into an operand, similarly as it is done for the installer in https://github.com/openshift/cluster-kube-apiserver-operator/blob/022057ed14a0f2b1eb98f7b5ccb76d100de011d6/pkg/operator/starter.go#L363):
+
+- [ ] test case to recover from a badly formatted or invalid pod manifest:
+
+  The bad manifest is rolled out, the API is unavailable, but recovers after 5min max by switching back to the old, save revision with reason NeverStartedUp. The operator does not retry, but stays degraded.
+
+- [ ] test case to recover from bad config observer output
+
+  The bad manifest is rolled out. The API is unavailable because the new kube-apiserver process does not start, but crash loops. I.e. the reason is CrashLooping. The operator does not retry, but stays degraded.
+
+- [ ] test case to recover from bad revisioned ConfigMaps and Secrets.
+
+  a) The manifest is rolled out, but a file in a configmap or secret (e.g. audit policy cannot be loaded. The API is unavailable because the new kube-apiserver process does not start, but crash loops. I.e. the reason is CrashLooping. The operator does not retry, but stays degraded.
+
+  b) TODO: think about a case where the ConfigMap/Secret exists, the apiserver starts up, does not get ready.
+
+- [ ] test case to recover from bad etcd smoke test
+
+  The new kube-apiserver starts up, but the etcd post-start hook never finishes. The reason is EtcdUnhealthy. The operator does retry and the cluster recovers when etcd is back again. For example wrong etcd endpoints can lead to this case.
+
+- [ ] test case to recover from non-etcd post-start hook never finishing
+
+  The new kube-apiserver starts up, but some non-etcd post-start hook never finishes. The reason is NotReady The operator does retry.
+
+### Graduation Criteria
+
+This will graduate directly to GA as 4.9 is the target release we have to hit.
+
+We might implement this enhancement partially in 4.9 and complete it in 4.10.
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, if that even matters for SNO pre-GA, the `last-known-good` symlink won't exist on first rollout. We will fall back to the previous revision then.
+
+On downgrade, the startup-monitor might still exist in `/etc/kubernetes`. We will make it wait until it sees an operand pod-manifest of the same revision before doing anything (see proposal).
```

Comment:
could the startup monitor exit when the pod revision is > than the current revision for N minutes?

Reply:
it would be restarted by kubelet.